mm2src/trading_api/src/one_inch_api/types.rs

```diff
@@ -0,0 +1,318 @@
+#![allow(clippy::result_large_err)]
+
+use super::client::QueryParams;
+use super::errors::ApiClientError;
+use ethereum_types::Address;
+use mm2_err_handle::mm_error::MmResult;
+use paste::paste;
+use serde::{Deserialize, Serialize};
+
+const ONE_INCH_MAX_SLIPPAGE: f32 = 50.0;
+const ONE_INCH_MAX_FEE_SHARE: f32 = 3.0;
+const ONE_INCH_MAX_GAS: u128 = 11500000;
+const ONE_INCH_MAX_PARTS: u32 = 100;
+const ONE_INCH_MAX_MAIN_ROUTE_PARTS: u32 = 50;
+const ONE_INCH_MAX_COMPLEXITY_LEVEL: u32 = 3;
+
+macro_rules! push_if_some {
+    ($arr: expr, $k: expr, $v: expr) => {
+        if let Some(v) = $v {
+            $arr.push(($k, v.to_string()))
+        }
+    };
+}
+
+macro_rules! def_with_opt_param {
+    ($var: ident, $var_type: ty) => {
+        paste! {
+            pub fn [<with_ $var>](mut self, $var: Option<$var_type>) -> Self {
+                self.$var = $var;
+                self
+            }
+        }
+    };
+}
```

Comment:
did you think about in-place modifications as it avoids moves and copies(if a case where the compiler needs/wants to), especially for large structs or frequent modifications?

```rust
pub fn [<with_ $var>](&mut self, $var: Option<$var_type>) -> &mut Self {
                 self.$var = $var;
                 self
             }
```

Reply:
I suggest to move `def_with_opt_param` macro to common.rs

Reply:
done in [23d8f1d](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2222/commits/23d8f1d6e387b45d6f91ad37f411bb08e976c368)